File: server/api/timeline.py
"""
时间轴API路由
"""
from fastapi import APIRouter, Depends, HTTPException
from typing import Dict, List
from collections import defaultdict
from server.schemas.timeline import TimelineResponse, TimelineItem, TimelineEvent, TimelineEmperor
from server.database.dependencies import get_db
from server.database.sqlite_manager import SQLiteManager

router = APIRouter()


@router.get("/{dynasty_id}", response_model=TimelineResponse)
async def get_timeline(
    dynasty_id: str,
    db: SQLiteManager = Depends(get_db)
):
    """获取指定朝代的时间轴数据"""
    try:
        # 获取朝代信息
        dynasty_sql = "SELECT dynasty_id, name, start_year, end_year FROM dynasties WHERE dynasty_id = ?"
        dynasty_row = db.fetch_one(dynasty_sql, (dynasty_id,))
        
        if not dynasty_row:
            raise HTTPException(status_code=404, detail=f"朝代不存在: {dynasty_id}")
        
        dynasty_name = dynasty_row[1]
        start_year = dynasty_row[2]
        end_year = dynasty_row[3]
        
        # 获取该朝代的所有事件
        events_sql = """
            SELECT event_id, title, event_type, start_date, location
            FROM events
            WHERE dynasty_id = ?
            ORDER BY start_date
        """
        event_rows = db.fetch_all(events_sql, (dynasty_id,))
        
        # 获取该朝代的所有皇帝
        emperors_sql = """
            SELECT emperor_id, name, temple_name, reign_start, reign_end
            FROM emperors
            WHERE dynasty_id = ?
            ORDER BY dynasty_order
        """
        emperor_rows = db.fetch_all(emperors_sql, (dynasty_id,))
        
        # 按年份组织时间轴数据
        timeline_dict: Dict[int, TimelineItem] = defaultdict(lambda: {
            "year": 0,
            "events": [],
            "emperor": None
        })
        
        # 添加事件到时间轴
        for event in event_rows:
            if event[3]:  # start_date
                try:
                    year = int(event[3][:4]) if isinstance(event[3], str) else event[3].year
                    timeline_dict[year]["year"] = year
                    timeline_dict[year]["events"].append({
                        "event_id": event[0],
                        "title": event[1],
                        "event_type": event[2],
                        "location": event[4]
                    })
                except (ValueError, AttributeError):
                    continue
        
        # 添加皇帝到时间轴（每年显示在位皇帝）
        for emperor in emperor_rows:
            if emperor[3] and emperor[4]:  # reign_start, reign_end
                try:
                    start = int(emperor[3][:4]) if isinstance(emperor[3], str) else emperor[3].year
                    end = int(emperor[4][:4]) if isinstance(emperor[4], str) else emperor[4].year
                    
                    for year in range(start, end + 1):
                        if year not in timeline_dict:
                            timeline_dict[year] = {
                                "year": year,
                                "events": [],
                                "emperor": None
                            }
                        
                        if timeline_dict[year]["emperor"] is None:
                            timeline_dict[year]["emperor"] = {
                                "emperor_id": emperor[0],
                                "name": emperor[1],
                                "temple_name": emperor[2],
                                "reign_start": emperor[3],
                                "reign_end": emperor[4]
                            }
                except (ValueError, AttributeError):
                    continue
        
        # 排序并转换为列表
        timeline = sorted(timeline_dict.values(), key=lambda x: x["year"])
        
        return {
            "dynasty_id": dynasty_id,
            "dynasty_name": dynasty_name,
            "start_year": start_year,
            "end_year": end_year,
            "timeline": timeline,
            "total_events": len(event_rows) if event_rows else 0,
            "total_emperors": len(emperor_rows) if emperor_rows else 0
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取时间轴数据失败: {str(e)}")

File: ios-app/HistoryGogo/Models/Timeline.swift
//
//  Timeline.swift
//  HistoryGogo
//
//  时间轴数据模型
//

import Foundation

/// 时间轴上的事件
struct TimelineEvent: Identifiable, Codable {
    let id: String
    let title: String
    let eventType: String
    let location: String?
    
    enum CodingKeys: String, CodingKey {
        case id = "event_id"
        case title
        case eventType = "event_type"
        case location
    }
}

/// 时间轴上的皇帝信息
struct TimelineEmperor: Codable {
    let emperorId: String
    let name: String
    let templeName: String?
    let reignStart: Date
    let reignEnd: Date?
    
    enum CodingKeys: String, CodingKey {
        case emperorId = "emperor_id"
        case name
        case templeName = "temple_name"
        case reignStart = "reign_start"
        case reignEnd = "reign_end"
    }
    
    var displayName: String {
        if let temple = templeName {
            return temple
        }
        return name
    }
}

/// 时间线上的单个时间点
struct TimelineItem: Identifiable, Codable {
    var id: Int { year }
    
    let year: Int
    let events: [TimelineEvent]
    let emperor: TimelineEmperor?
    
    /// 该年是否有重要事件
    var hasEvents: Bool {
        return !events.isEmpty
    }
    
    /// 事件数量
    var eventCount: Int {
        return events.count
    }
}

/// 完整的时间轴响应
struct TimelineResponse: Codable {
    let dynastyId: String
    let dynastyName: String
    let startYear: Int
    let endYear: Int
    let timeline: [TimelineItem]
    let totalEvents: Int
    let totalEmperors: Int
    
    enum CodingKeys: String, CodingKey {
        case dynastyId = "dynasty_id"
        case dynastyName = "dynasty_name"
        case startYear = "start_year"
        case endYear = "end_year"
        case timeline
        case totalEvents = "total_events"
        case totalEmperors = "total_emperors"
    }
    
    /// 时间跨度（年）
    var duration: Int {
        return endYear - startYear
    }
    
    /// 按年份获取时间轴项
    func item(for year: Int) -> TimelineItem? {
        return timeline.first { $0.year == year }
    }
}

// MARK: - Sample Data

extension TimelineEvent {
    static let sample = TimelineEvent(
        id: "event_001",
        title: "明朝建立",
        eventType: "政治",
        location: "南京"
    )
}

extension TimelineEmperor {
    static let sample = TimelineEmperor(
        emperorId: "ming_taizu",
        name: "朱元璋",
        templeName: "明太祖",
        reignStart: Date(timeIntervalSince1970: -19000000000),
        reignEnd: Date(timeIntervalSince1970: -18000000000)
    )
}

extension TimelineItem {
    static let sample = TimelineItem(
        year: 1368,
        events: [
            TimelineEvent(id: "event_001", title: "明朝建立", eventType: "政治", location: "南京")
        ],
        emperor: TimelineEmperor.sample
    )
    
    static let samples: [TimelineItem] = [
        TimelineItem(
            year: 1368,
            events: [
                TimelineEvent(id: "event_001", title: "明朝建立", eventType: "政治", location: "南京")
            ],
            emperor: TimelineEmperor.sample
        ),
        TimelineItem(
            year: 1369,
            events: [],
            emperor: TimelineEmperor.sample
        ),
        TimelineItem(
            year: 1370,
            events: [
                TimelineEvent(id: "event_002", title: "北伐蒙古", eventType: "军事", location: "北方")
            ],
            emperor: TimelineEmperor.sample
        )
    ]
}

extension TimelineResponse {
    static let sample = TimelineResponse(
        dynastyId: "ming",
        dynastyName: "明朝",
        startYear: 1368,
        endYear: 1644,
        timeline: TimelineItem.samples,
        totalEvents: 150,
        totalEmperors: 16
    )
}
